design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,219 @@
+
+## Motivation
+Running High performance VMs with an even number of CPUs with the conditions below.
+
+### The problem
+In order to replicate this issue, the following conditions must be met:
+* The node has hyperthreading (SMT) enabled.
+* Kubelet’s CPUManager is set to static - full-pcpus-only.
+* Kubevirt CR's defaultRuntimeClass is set to a high performance runtimeClass.
+* VM is set with `dedicatedCPUPlacement` and `IsolateEmulatorThread` enabled.
+* VM requests an even number of CPUs.
+
+When enabling `IsolateEmulatorThread` on a VM with even CPU count, Kubevirt is requesting one extra CPU for the virt-launcher pod. This is an internal change that is not exposed to the user.
+After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
+For example, let’s consider this VM spec:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+kubelet will emit following event:
+```
+SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
+```
+The following picture helps show how this is a problem
+<img src="smt-slignment-error-diagram.png">
+
+## Goals
+Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.
+
+## Non Goals
+Set the virt-launcher Pod scheduling to match a node with SMT aligned to its request.
+
+## Definition of Users
+This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+* As a cluster admin, I want to enable VMs with an even number of CPUs and `isolatedEmulatorThread` enabled to run on my cluster by setting a cluster-wide knob on HCO.
+* As a Kubevirt VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadCompleteToEvenParity`:
+```
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  annotations:
+    kubevirt.io/EmulatorThreadCompleteToEvenParity:
+spec:
+  configuration:
+    developerConfiguration:
+      featureGates:
```

Comment:
Having a feature gate to protect non-API (an annotation) seems redundant and arbitrary. Not having a feature gate would also allow you to remove the webhook checking that a VM annotation is only set if the CR annotation is set. Please consider removing it.

Reply:
That's a practice commonly practiced on kubevirt (for example [IgnitionEnabled](https://github.com/kubevirt/kubevirt/blob/7e3aa9fff0226fea881fa4efc7906bad3cfcf4be/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go#L1682)) and was specifically requested by @xpivarc.
@xpivarc - Do you think we can let go of the FG check from the create-admitter validation webhook?